### DISCUSSION TITLE
How to display a Context Menu on a Mobile Device

### DISCUSSION THREAD

**Discussion by swordream** [dotnet/maui#10752] [Q&A]
With the .NET MAUI in .NET 7 Release Candidate 1,We can now attach a context menu to any visual element using a MenuFlyout control. When the user right-clicks that view, the flyout will appear in that location when run on a desktop platform.
 
So,why not apply this feature to mobile device,for example,long tap the item to invoke the context menu, i think it's very useful。

**Comment by nebula2** [comment]
on mobile you typically use a swipeview

https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/swipeview

**Reply by hansmbakker** [reply]
That depends on the action and amount of actions - if you want to display more actions than two, you still need a different solution.

For example Try press & holding on a contact in the contacts app on iOS or Android. Both have implemented it as a context menu.
![image](https://github.com/dotnet/maui/assets/3463496/bbf58bce-aa95-48bc-a199-2a58850b2025)
![image](https://github.com/dotnet/maui/assets/3463496/1f48b02d-4c60-4669-a142-5e95bf34b9da)

Apple actually uses long press to open a context menu on iOS a lot.

An alternative could be implementing it as a bottomsheet, but that was not the ask.

**Comment by ghost** [comment]
We've moved this issue to the Backlog milestone. This means that it is not going to be worked on for the coming release. We will reassess the backlog following the current release and consider this item at that time. To learn more about our issue management process and to have better expectation regarding different types of issues you can read our [Triage Process](https://github.com/dotnet/maui/blob/main/docs/TriageProcess.md).

**Comment by swordream** [comment]
> on mobile you typically use a swipeview
> 
> https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/swipeview

Yes, but sometimes the context menu is also very useful. WeChat uses a lot of context menus instead of the previous SwipeView。

**Comment by knocte** [comment]
> why not apply this feature to mobile device,for example,long tap the item to invoke the context menu, i think it's very useful

Agreed.

**Reply by kcoop** [reply]
Also agreed. Especially for rich canvas-oriented apps. Swipe view is only useful for lists.

**Comment by anpin** [comment]
One can use this package https://github.com/anpin/ContextMenuContainer (shameless plug!). Which was just updated to support MAUI. Allows one to display context menu with long-tap on mobile and right-click on desktop.

**Reply by hansmbakker** [reply]
@anpin That looks helpful, that it also supports desktop.
Do you have experience with https://github.com/the49ltd/The49.Maui.ContextMenu? Do you have a comparison with your library?

**Reply by daveikaye2** [reply]
I cannot seem to get your NuGet package to work (.NET 8, CommunityToolkit.Maui=9.0.3, Microsoft.Maui.Controls=8.0.82, Microsoft.Maui.Controls.Compatibility=8.0.82)...  I need it to work in WinUI and iOS.

Trying in WinUI:
+Added ContextMenuContainer v1.2.3 in my proj file
+Added APES.UI.XF.ContextMenuContainer.Init() in the constructor of App.xaml.cs (root of the project)
+Added `xmlns:c="http://apes.ge"` and 
```
<StackLayout>
    <c:ContextMenuContainer x:Name="container1">
        <c:ContextMenuContainer.MenuItems>
            <c:ContextMenuItem Text="Action 1" />
            <c:ContextMenuItem Text="Action with icon" />
        </c:ContextMenuContainer.MenuItems>
        <c:ContextMenuContainer.Content>
            <Frame BackgroundColor="ForestGreen" Padding="24" CornerRadius="0" >

                <Label Text="***FOO FOO!!!" HorizontalTextAlignment="Center" TextColor="White" FontSize="36" HorizontalOptions="FillAndExpand" VerticalOptions="Center" MinimumHeightRequest="30" MinimumWidthRequest="100"/>

            </Frame>
        </c:ContextMenuContainer.Content>
    </c:ContextMenuContainer>
</StackLayout>
```
to the XAML where I want to have the context menu.  Everything compiles.

I see the ***FOO FOO!!! being rendered on the nice green background, but no context menu is popping up.  I left-click, left-click hold, right-click, right-click hold... nothing.

Any ideas?

**Reply by emayevski** [reply]
> One can use this package https://github.com/anpin/ContextMenuContainer (shameless plug!). Which was just updated to support MAUI. Allows one to display context menu with long-tap on mobile and right-click on desktop.

It would be a great addition, if it worked in MAUI. I tried to build it for MAUI 9 but gave up - too many problems with MAUI. 
Additionally, I could not figure out from the current code, how it even pops up. Could it be that some code was lost, as the previous poster reported the issue of the menu not appearing as well?

**Reply by anpin** [reply]
Please try updated version published yesterday. Support for Xamarin.Forms was removed and sample using latest MAUI is now tested in github actions.

**Comment by anpin** [comment]
I don’t have a comparison, but I’ve been using it four couple years since the Xamarin.Forms was around. The delay in porting the lib over to MAUI was due to various issues preventing my apps from being effectively shipped which has been mostly resolved by now. Please let me know if you find it helpful!   On Sun, Feb 11, 2024 at 11:08 PM, Hans Bakker ***@***.***> wrote:  
@anpin That looks helpful, that it also supports desktop.
Do you have experience with https://github.com/the49ltd/The49.Maui.ContextMenu? Do you have a comparison with your library?

**Comment by anpin** [comment]
Please open an issue in the repo.

**Reply by bohdan-bats-bs** [reply]
I opened it. You can see it [here](https://github.com/dotnet/maui/issues/33402#issuecomment-3725599419)

**Comment by Kiriakos41** [comment]
+1

**Comment by bohdan-bats-bs** [comment]
+1

**Comment by bohdan-bats-bs** [comment]
Very important feature as context menu is used very often. For mobile devices it should appear on long tap on visual element

**Comment by bohdan-bats-bs** [comment]
I suggest to add ContextMenu property to VisualElement like it is in WPF

**Comment by kirquebe** [comment]
+1, seems pretty common-sense to me

**Reply by pictos** [reply]
Right now you can use my library for that:
- https://www.nuget.org/packages/PJSouzaSoftware.ContextActions.Maui

**Comment by pictos** [comment]
For those who are blocked, you can use my library for now.

- https://www.nuget.org/packages/PJSouzaSoftware.ContextActions.Maui

**Reply by RedEye-Developers** [reply]
thanks for this lib.

it is possible to change the ActionBehavior from Hold to Click to open contextMenu in android ?

**Reply by bohdan-bats-bs** [reply]
> thanks for this lib.
> 
> it is possible to change the ActionBehavior from Hold to Click to open contextMenu in android ?

For that you can use this library: https://github.com/DIPSAS/DIPS.Mobile.UI/wiki/Context-Menus

**Reply by RedEye-Developers** [reply]
thanks, it is working perfect what i expected.
